solver/src/settlement_submission/archer_settlement.rs

```diff
@@ -0,0 +1,377 @@
+// Design:
+// As in the traditional transaction submission workflow the main work in this module is checking
+// the gas price in a loop and updating the transaction when the gas price increases. This differs
+// so that we can make use of the property that archer transactions do not cost cast if they fail.
+// When we detect that the transaction would no longer succeed we stop trying to submit and return
+// so that the solver can run again.
+// In addition to simulation failure we make use of a deadline after which submission attempts also
+// stop. This allows the solver to update and improve a solution even if it hasn't yet become
+// invalid.
+// We do not know in advance which of our submitted transactions will get mined. Instead of polling
+// all of them we only check the account's nonce as an optimization. When this happens all our
+// transactions definitely become invalid (even if the transaction came for whatever reason
+// from outside) so it is only at that point that we need to check the hashes individually to the
+// find the one that got mined (if any).
+
+// Idea:
+// The current driver code is either solving or waiting for a solution to be mined.
+// When using archer we can improve on this by always solving the current order book as if there was
+// no pending transaction and continually updating the transaction we are sending to archer.
+// This is a bigger change so the code here still adheres to the traditional tx submission
+// workflow.
+
+// TODO: Make sure settlement contract always has eth to pay the miner. Could reuse buffers or just
+// manually fund it like we do with other solver accounts and make part of our monitor scripts.
+
+// TODO: Node gas estimates seem a little higher than actual. This can be seen when comparing
+// to tenderly simulations and mined transactions. This causes us to overpay somewhat.
```

Comment:
I think this is because the tipping contract goes from 0 balance -> non-zero balance which has a higher gas costs from going from non-zero balance A -> non-zero balance B. Then, a portion of that gas may get refunded because it goes back to its initial state by the end of the transaction.

It works like this for storage, so I'm *assuming* that there may be something similar at play here, but I'm not 100% sure.